src/track/beats.cpp

```diff
@@ -37,20 +534,376 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+        return std::unique_ptr<BeatsInternal::iterator>();
+    }
+
+    Beat startBeat(startFrame), stopBeat(stopFrame);
+
+    BeatList::const_iterator firstBeat =
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), startBeat);
+
+    BeatList::const_iterator lastBeat =
+            std::upper_bound(m_beats.cbegin(), m_beats.cend(), stopBeat);
+    if (lastBeat >= m_beats.cbegin()) {
+        lastBeat = m_beats.cend() - 1;
+    }
+
+    if (firstBeat >= lastBeat) {
+        return std::unique_ptr<BeatsInternal::iterator>();
+    }
+    return std::make_unique<BeatsInternal::iterator>(firstBeat, lastBeat + 1);
+}
+bool BeatsInternal::hasBeatInRange(
+        FramePos startFrame, FramePos stopFrame) const {
+    if (!isValid() || startFrame > stopFrame) {
+        return false;
+    }
+    FramePos curBeat = findNextBeat(startFrame);
+    if (curBeat <= stopFrame) {
+        return true;
+    }
+    return false;
+}
+FramePos BeatsInternal::findNextBeat(FramePos frame) const {
+    return findNthBeat(frame, 1);
+}
+FramePos BeatsInternal::findPrevBeat(FramePos frame) const {
+    return findNthBeat(frame, -1);
+}
+double BeatsInternal::getBpmRange(
+        FramePos startFrame, FramePos stopFrame) const {
+    if (!isValid()) {
+        return -1;
+    }
+    Beat startBeat(startFrame), stopBeat(stopFrame);
+    return calculateBpm(startBeat, stopBeat).getValue();
+}
+Bpm BeatsInternal::getBpmAroundPosition(FramePos curFrame, int n) const {
+    if (!isValid()) {
+        return Bpm();
+    }
+
+    // To make sure we are always counting n beats, iterate backward to the
+    // lower bound, then iterate forward from there to the upper bound.
+    // kInvalidFramePos indicates we went off the map -- count from the beginning.
+    FramePos lower_bound = findNthBeat(curFrame, -n);
+    if (lower_bound == kInvalidFramePos) {
+        lower_bound = m_beats.first().getFramePosition();
+    }
+
+    // If we hit the end of the beat map, recalculate the lower bound.
+    FramePos upper_bound = findNthBeat(lower_bound, n * 2);
+    if (upper_bound == kInvalidFramePos) {
+        upper_bound = m_beats.last().getFramePosition();
+        lower_bound = findNthBeat(upper_bound, n * -2);
+        // Super edge-case -- the track doesn't have n beats!  Do the best
+        // we can.
+        if (lower_bound == kInvalidFramePos) {
+            lower_bound = m_beats.first().getFramePosition();
+        }
+    }
+
+    // TODO(JVC) We are extracting frame numbers to then construct beats.
+    // Then in calculateBpm we are using the frame position to find
+    // the beats to  use them to calculate. Seems inefficient
+    // Will not make more sense to extract the Beats straight?
+    // We can use getBpmRange and move the logic of calculateBpm there
+    Beat startBeat(lower_bound), stopBeat(upper_bound);
+    return calculateBpm(startBeat, stopBeat);
+}
+
+void BeatsInternal::setSignature(TimeSignature sig, int downbeatIndex) {
```

Comment:
This does not work how I expect. I think changing the note value should not add or remove any downbeats. Instead, it should insert new beats half way between each beat to double the note value or remove every other beat to half the note value.